### PR TITLE
Fluka materials

### PR DESCRIPTION
**Pull request by lequ0n**

### CODE REVIEW

**Comment by lequ0n** [comment]
In Card class 'tag' was changed to 'codewd' to match description in FLUKA https://flukafiles.web.cern.ch/manual/chapters/general_features_of_fluka_input.html

**Comment by lequ0n** [comment]
The UUID of materials is far too long, while for individual materials we can rely on the number to be assigned by FLUKA, for compounds the name is required. I have decided to create random 10 letter names for materials and substances, if there is a better solution I am happy to consider it

**Comment by lequ0n** [comment]
Currently predefined materials and compounds are loaded from json file. Opening file looks like this: `Path("__file__").resolve().parent / "converter" / "fluka" / "predefined_materials.json"`. Is there alternative method to do this or should I rewrite it as a dataclass in python file?

**Comment by lequ0n** [comment]
In current state if density of material/compound is changed, then new material/compound with random 8 letter name is created. To ensure consistent results (to pass tests) seed has been set to 42.

**Comment by lequ0n** [comment]
Tests to the new code are currently missing and ionisation override and custom stopping power are not yet implemented.

File: converter/fluka/cards/assignmat_card.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from converter.fluka.cards.card import Card
+from converter.fluka.helper_parsers.material_parser import MaterialAssignment
+
+
+@dataclass
+class AssignmatsCard:
+    """
+    Class representing description of material assignment in FLUKA input.
+    Every material assignment is represented by one line:
+        codwed - "ASSIGNMA"
```
Comment (grzanka) [inline converter/fluka/cards/assignmat_card.py:11]: what is `codewd`  ? "code word" ?

Comment (lequ0n) [reply converter/fluka/cards/assignmat_card.py:11]: I've changed it to match description in FLUKA:
https://flukafiles.web.cern.ch/manual/chapters/general_features_of_fluka_input.html

**Comment by lequ0n** [comment]
UI supports elements up to an atomic number of 98, FLUKA can handle Einsteinium and Fermium in addition, but I have limited the converter to 98 only.
As of compounds, unfortunately FLUKA only supports 11 predefined ones, if we want additional ones we have to define them.
Regarding ionisation potential, I'm not sure if this is supported by UI, and if so, I don't know where exactly. I put a placeholder in the  code.
As for the tests my first thought was to do the current ones. I'm not 100% convinced that they are good enough.
The file HASN'T been run in FLUKA yet, I need to run it before merging to main

**Comment by grzanka** [comment]
> UI supports elements up to an atomic number of 98, FLUKA can handle Einsteinium and Fermium in addition, but I have limited the converter to 98 only. 

We can safely ifnore Einsteinium and Fermium, nobody uses them.

> As of compounds, unfortunately FLUKA only supports 11 predefined ones, if we want additional ones we have to define them. 

At this moment lets start with 11 predefined

> Regarding ionisation potential, I'm not sure if this is supported by UI, and if so, I don't know where exactly. 

UI doesn't support it yet, but you can foresee a field in JSON for that. In converter you can check if ionisation potential is present. If that is the case (its present) then use it in converter.

> I put a placeholder in the code. As for the tests my first thought was to do the current ones. I'm not 100% convinced that they are good enough. The file HASN'T been run in FLUKA yet, I need to run it before merging to main

Do a test with running with Fluka !

**Comment by lequ0n** [comment]
Running the file returned by converter made me realize that defining a MATERIAL card with density and atomic number is not enough, it is also necessary to add a LOW-MAT card. Adding LOW-MAT limits the number of molecules supported. ICRUs that are in the UI and not supported by the converter { 34, 37, 44, 45, 52, 59, 61, 66-70, 76, 81, 84-88, 90, 93, 96, 97, 98 }
Simulations run successfully in FLUKA, although it is hard for me to determine if the results are good.

**Comment by grzanka** [comment]
@lequ0n are you planning to add something here or can we proceed with review ?

**Comment by lequ0n** [comment]
> @lequ0n are you planning to add something here or can we proceed with review ?

We can proceed with review (I forgot to mention on discord), I've fixed material assignment and now it runs on Fluka.

**Review by grzanka — APPROVED** [review]

**Pull request merged**